proof/refine/ARM/Invariants_H.thy

```diff
@@ -250,15 +250,23 @@ abbreviation scs_of' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> sched_
 abbreviation scReplies_of :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
   "scReplies_of s \<equiv> scs_of' s |> scReply"
 
-definition tcb_of' :: "kernel_object \<Rightarrow> tcb option" where
-  "tcb_of' kobj \<equiv> (case kobj of KOTCB tcb \<Rightarrow> Some tcb
-                               | _         \<Rightarrow> None)"
+abbreviation tcb_of' :: "kernel_object \<Rightarrow> tcb option" where
+  "tcb_of' \<equiv> projectKO_opt"
 
 abbreviation tcbs_of' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> tcb option" where
   "tcbs_of' s \<equiv> ksPSpace s |> tcb_of'"
 
-abbreviation tcb_scs_of' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
-  "tcb_scs_of' s \<equiv> tcbs_of' s |> tcbSchedContext"
+abbreviation tcbSCs_of :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
+  "tcbSCs_of s \<equiv> tcbs_of' s |> tcbSchedContext"
+
+abbreviation scTCBs_of :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
+  "scTCBs_of s \<equiv> scs_of' s |> scTCB"
+
+abbreviation scTCB_sym_refs where
+  "scTCB_sym_refs s \<equiv> sym_heap (tcbSCs_of s) (scTCBs_of s)"
+
+abbreviation scReply_sym_refs where
+  "scReply_sym_refs s \<equiv> sym_heap (scReplies_of s) (replySCs_of s)"
```

Comment:
Cheers.

I like having the abbreviation because it means I don't need to write out all the lambdas and s's and parentheses. And on all other points I agree, though I prefer the less verbose version (2-2).